src/main/java/com/amazon/ion/impl/_Private_IonReaderBuilder.java

```diff
@@ -272,31 +296,60 @@ interface IonReaderFromInputStreamFactoryBinary {
         IonReader makeReader(_Private_IonReaderBuilder builder, InputStream source, byte[] alreadyRead, int alreadyReadOff, int alreadyReadLen);
     }
 
+    /**
+     * Reads from the given source into the given byte array, stopping once either
+     * <ol>
+     *     <li>`length` bytes have been read, or</li>
+     *     <li>the end of the source stream has been reached, or</li>
+     *     <li>the source stream throws an exception.</li>
+     * </ol>
+     * @param source the source of the bytes to read.
+     * @param destination the destination for the bytes read.
+     * @param length the number of bytes to attempt to read.
+     * @return the number of bytes read into `destination`.
+     */
+    private static int fillToLengthOrStreamEnd(InputStream source, byte[] destination, int length) {
+        int bytesRead = 0;
+        while (bytesRead < length) {
+            int bytesToRead = length - bytesRead;
+            int bytesReadThisIteration;
+            try {
+                bytesReadThisIteration = source.read(destination, bytesRead, bytesToRead);
+            } catch (IOException e) {
+                // Some InputStream implementations throw IOExceptions (e.g. EOFException) in certain cases to convey
+                // that the end of the stream has been reached.
+                break;
+            }
```

Comment:
I think this `catch` should be removed.

Throwing an exception at EOF violates the [contract](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/io/InputStream.html#read(byte%5B%5D,int,int)) of `read`, which specifically states (emphasis added):

> [IOException](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/io/IOException.html) - If the first byte cannot be read for any reason **other than end of file**, or if the input stream has been closed, or if some other I/O error occurs.

If the exception is _not_ expressing EOF, then this code swallows a significant exception indicating something is broken with the stream.

Reply:
Yes, agreed, the general IOException case should throw. We will catch and handle `EOFException` instead, as this is known to be used by GZIPInputStream. See this test for an illustration: https://github.com/amazon-ion/ion-java/blob/master/src/test/java/com/amazon/ion/system/IonReaderBuilderTest.java#L205

This fails if we don't handle EOFException here.